go/ephemeral/team_ek.go

```diff
@@ -0,0 +1,413 @@
+		}
+	}
+
+	// If we didn't short circuit above, then the signing key is correct. Parse
+	// the JSON and return the result.
+	parsedStatement := keybase1.TeamEkStatement{}
+	err = json.Unmarshal(payload, &parsedStatement)
+	if err != nil {
+		return nil, false, err
+	}
+	return &parsedStatement, false, nil
+}
+
+func filterStaleTeamEKStatement(ctx context.Context, g *libkb.GlobalContext, statement *keybase1.TeamEkStatement, merkleRoot libkb.MerkleRoot) (active []keybase1.TeamEkMetadata, err error) {
+	defer g.CTrace(ctx, "filterStaleTeamEKStatement", func() error { return err })()
+
+	if statement == nil {
+		return nil, err
+	}
+
+	allMetadata := append([]keybase1.TeamEkMetadata{}, statement.ExistingTeamEkMetadata...)
+	allMetadata = append(allMetadata, statement.CurrentTeamEkMetadata)
+	for _, metadata := range allMetadata {
+		if !ctimeIsStale(metadata.Ctime, merkleRoot) {
+			active = append(active, metadata)
+		}
+	}
+
+	return active, nil
+}
+
+type teamMemberEKStatementResponse struct {
+	Sigs map[keybase1.UID]string `json:"sigs"`
+}
+
+// Returns nil if all team members have never published a teamEK. Verifies that
+// the map of users the server returns are indeed valid team members of the
+// team and all signatures verify correctly for the users.
+func fetchTeamMemberStatements(ctx context.Context, g *libkb.GlobalContext, teamID keybase1.TeamID) (statementMap map[keybase1.UID]*keybase1.UserEkStatement, err error) {
+	defer g.CTrace(ctx, "fetchTeamMemberStatements", func() error { return err })()
+
+	apiArg := libkb.APIArg{
+		Endpoint:    "team/member_eks",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args: libkb.HTTPArgs{
+			"team_id": libkb.S{Val: string(teamID)},
+		},
+	}
+	res, err := g.GetAPI().Get(apiArg)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedResponse := teamMemberEKStatementResponse{}
+	err = res.Body.UnmarshalAgain(&parsedResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	team, err := teams.Load(ctx, g, keybase1.LoadTeamArg{
+		ID: teamID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	statementMap = make(map[keybase1.UID]*keybase1.UserEkStatement)
+	for uid, sig := range parsedResponse.Sigs {
+		// Verify the server only returns actual members of our team.
+		uv, err := team.UserVersionByUID(ctx, uid)
+		if err != nil {
+			return nil, err
+		}
+		isMember := team.IsMember(ctx, uv)
+		if !isMember {
+			g.Log.CWarningf(ctx, "Server lied about team membership! %v is not a member of team %v", uv, teamID)
+			return nil, err
+		}
+		memberStatement, wrongKID, err := verifySigWithLatestPUK(ctx, g, uid, sig)
+		// Check the wrongKID condition before checking the error, since an error
+		// is still returned in this case. TODO: Turn this warning into an error
+		// after EK support is sufficiently widespread.
+		if wrongKID {
+			g.Log.CWarningf(ctx, "It looks like you revoked a team member without generating new ephemeral keys. Are you running an old version?")
```

Comment:
I don't think revoking a team member can cause this condition. I think it's the same condition that it was in the userEK case: Someone has rolled their PUK without issuing a new userEK. It's in verifying the sig on a TeamEKStatement that we might run into the team-level version of this problem.

Reply:
oops yeah these kinda got lost in translation during the refactors..